frontend/Dockerfile

```diff
@@ -1,11 +1,46 @@
-FROM node:18-alpine
+# Install dependencies only when needed
+FROM node:18-alpine AS deps
 
-COPY . /app
 WORKDIR /app
 
-RUN cd /app \
-    && yarn \
-    && yarn next telemetry disable \
-    && yarn next build
+COPY package.json yarn.lock ./
 
-CMD yarn start
+RUN yarn
+
+# Rebuild the source code only when needed
+FROM node:18-alpine AS builder
+
+WORKDIR /app
+
+COPY . .
+COPY --from=deps /app/node_modules ./node_modules
+
+RUN NEXT_PUBLIC_API_BASE_URL="APP_PLACEHOLDER_NEXT_PUBLIC_API_BASE_URL" NEXT_PUBLIC_HCAPTCHA_SITE_KEY="APP_PLACEHOLDER_NEXT_PUBLIC_HCAPTCHA_SITE_KEY" yarn build
+
+# Production image, copy all the files and run next
+FROM node:18-alpine AS runner
+
+WORKDIR /app
+
+ENV NODE_ENV production
+
+COPY --from=builder /app/public ./public
+COPY --from=builder /app/.next ./.next
+COPY --from=builder /app/node_modules ./node_modules
+COPY --from=builder /app/package.json ./package.json
+COPY --from=builder /app/docker-entrypoint.sh ./entrypoint.sh
+
+RUN addgroup -g 1001 -S nodejs && \
+    adduser -S nextjs -u 1001 -G nodejs && \
+    chown -R nextjs:nodejs /app/.next && \
+    chmod +x /app/entrypoint.sh
```

Comment:
I personally prefer to write out the arguments as full names, e.g. `--system` instead of `-S`, because the explicitness outweighs the conciseness IMO, but this is fine.

Reply:
I forgot what the -S flag did, and the linked examples had them so I just left them as is. I'll change this all tomorrow